### DISCUSSION TITLE
ICMP packets aren't sent

### DISCUSSION THREAD

**Discussion by tk-nguyen** [libpnet/libpnet#512] [Q&A] [answered]
Hello, I'm writing a simple ICMP program to test the functionality of this crate. When I try to run it, I can't see any packets sent out. I've tried capturing packets on my interface, but it came up with nothing. Here's my code:
```rust
use pnet::packet::icmp::{echo_request::MutableEchoRequestPacket, IcmpTypes};
use pnet::packet::ip::IpNextHeaderProtocols::Icmp;
use pnet::transport::{icmp_packet_iter, transport_channel, TransportChannelType};
use std::net::{IpAddr, Ipv4Addr};

fn main() {
    let (mut tx, mut rx) = match transport_channel(1500, TransportChannelType::Layer3(Icmp)) {
        Ok((tx, rx)) => (tx, rx),
        Err(e) => panic!("An error occurred when creating transport channel: {}", e),
    };
    let mut payload = [0u8; 64];
    let dst = "192.168.100.101".parse::<Ipv4Addr>().unwrap();
    let mut recv = icmp_packet_iter(&mut rx);
    loop {
        let mut ping = MutableEchoRequestPacket::new(&mut payload).unwrap();
        ping.set_icmp_type(IcmpTypes::EchoRequest);
        ping.set_payload(b"hello");
        println!("Bytes sent: {}", tx.send_to(ping, IpAddr::V4(dst)).unwrap());
        match recv.next() {
            Ok((pkt, addr)) => println!("{:#?} {}", pkt, addr),
            Err(e) => eprintln!("Error: {} ", e),
        }
    }
}
```
And here's the output of tcpdump:
![image](https://user-images.githubusercontent.com/39090621/124798169-423afc00-df7d-11eb-83f5-148374f91b48.png)
I use sudo because `transport_channel` won't run under normal user permissions.
Any help would be appreciated!

**Comment by stappersg** [comment]
On Wed, Jul 07, 2021 at 09:45:28AM -0700, Thai Nguyen wrote:
> Any help would be appreciated!

Which network interface should the programm use?

**Reply by tk-nguyen** [reply]
I don't think I need to specify one? The [`transport_echo_server`](https://github.com/libpnet/libpnet/blob/master/examples/transport_echo_server.rs) example doesn't say anything about interfaces. Unless I'm missing something here.

**Comment by stappersg** [comment, marked as answer]
On Wed, Jul 07, 2021 at 09:31:48PM -0700, Thai Nguyen wrote:
> Unless I'm missing something here.

use pnet::transport::TransportChannelType::Layer4;
  versus
use pnet::transport::{icmp_packet_iter, transport_channel, TransportChannelType};

**Reply by tk-nguyen** [reply]
After changing the `TransportChannelType` from `Layer3` to `Layer4`, it works perfectly! Thank you for your help!
